### ISSUE TITLE
Wrapped python packages don't propagate their PYTHONPATH

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nixpkgs#23676]
Related to #23639 but not quite the same.

If a python package launches python, the child's `PYTHONPATH` will know nothing of the dependencies of the parent. This is almost never what you want. I think this is because rather than setting `PYTHONPATH`, we use `site` inside python to set the path for the current interpreter. Spawning new child interpreters won't get any of that stuff.

My current inclination would be to add a very simple line to the wrapper shim we have today:

```python
os.environ['PYTHONPATH'] = ":".join(sys.path)
```

Right after we call `reduce` to add all the paths. That would ensure that all children inherit the right environment. Any thoughts?

cc @FRidh

**Comment by FRidh** [comment]
Originally we used `--prefix PYTHONPATH` in which case child processes do inherit `PYTHONPATH` but when one had several programs with different `PYTHONPATH` calling eachother it caused problems. Therefore, @abbradar proposed using `--set PYTHONPATH` which solved that issue, but I was against it because a) it breaks usage of a Python feature, `PYTHONPATH`, and b) certain applications couldn't be used anymore (like `ipython`) and it caused problems with the `installCheckPhase`.

Several days ago I tested another solution where I added a `sitecustomize.py` to the interpreter. That file is automatically loaded by the interpreter and in it I let it check for `NIX_PYTHONPATH` and then add that to `site` or `sys.path`. Then, we could use `--set NIX_PYTHONPATH` without it breaking usage of `PYTHONPATH`. Throughout Nixpkgs we should then use `NIX_PYTHONPATH` instead of `PYTHONPATH`.

**Comment by 0xABAB** [comment]
@FRidh I like your partial analysis of the problem space. One way to make progress on this is to write down all the use cases (and not some subset, however appealing that is for people who have a problem that needs to be fixed "now") and only after there is agreement on that, generate candidate solutions for those, and then finally implement one of those. 

One output of the use cases phase would be end-user documentation specifying what is possible. There should also be integration tests showing that these actually work (and linked from the documentation).

Doing it in the above way would make sure that at some point the Python integration is actually completed, as opposed to merely improved.

I am not against incremental temporary hacks which do not destroy the possibility of another proper design later, but it should not introduce distributed assumptions, which a good design wouldn't have.

**Comment by FRidh** [comment]
@0xABAB I agree that we should put this on paper. It has taken me quite a lot of time to get a good understanding of Python packaging in general, and how we deal with it on Nix. A while ago I [wrote down a bit](https://gist.github.com/FRidh/c3e974e34eac738405af699786b462d5). It's quite brief and already partially outdated so I've now begun updating it. It's very much work in progress :)

**Comment by mmahut** [comment]
Any news on this issue?

**Comment by FRidh** [comment]
Subprocess should not inherit `PYTHONPATH` because it will cause breakage when other Python applications inherit `PYTHONPATH`. We now have `NIX_PYTHONPATH` in a `sitecustomize.py` that is automatically unset, thus preventing leaking. This is currently only used in `buildEnv`. We should use this eventually elsewhere as well.

Related tickets:
- #16591
- #11423 (which kind of requests the opposite)
